File: library/src/main/java/com/thesurix/gesturerecycler/transactions/RemoveTransaction.java
package com.thesurix.gesturerecycler.transactions;


import com.thesurix.gesturerecycler.GestureAdapter;
import com.thesurix.gesturerecycler.GestureViewHolder;

/**
 * @author thesurix
 */
public class RemoveTransaction<T> implements AdapterTransaction {

    private final GestureAdapter<T, ? extends GestureViewHolder> mAdapter;
    private T mItem;
    private final int mPosition;

    public RemoveTransaction(final GestureAdapter<T, ? extends GestureViewHolder> adapter, final int position) {
        mAdapter = adapter;
        mPosition = position;
    }

    @Override
    public boolean perform() {
        mItem = mAdapter.getData().remove(mPosition);
        final boolean success = mItem != null;
        if (success) {
            mAdapter.notifyItemRemoved(mPosition);
        }
        return success;
    }

    @Override
    public boolean revert() {
        if (mItem != null) {
            mAdapter.getData().add(mPosition, mItem);
            mAdapter.notifyItemInserted(mPosition);
            return true;
        }

        return false;
    }
}

File: library/src/main/java/com/thesurix/gesturerecycler/transactions/RemoveTransaction.kt
package com.thesurix.gesturerecycler.transactions


/**
 * @author thesurix
 */
class RemoveTransaction<T>(private val position: Int,
                           private val headerEnabled: Boolean) : Transaction<T> {
    private var item: T? = null

    override fun perform(transactional: Transactional<T>): Boolean {
        return with(transactional.data) {
            val removedItem = removeAt(position)
            removedItem?.let {
                item = it
                transactional.notifyRemoved(position + if(headerEnabled) 1 else 0)
                true
            } ?: false
        }
    }

    override fun revert(transactional: Transactional<T>): Boolean {
        return with(transactional.data) {
            item?.let {
                add(position, it)
                transactional.notifyInserted(position + if(headerEnabled) 1 else 0)
                true
            } ?: false
        }
    }
}
